### PR TITLE
CASTER layer implementation

### PR DESCRIPTION
**Pull request by andriy-nikolov**
# Summary
 
An implementation of the CASTER model layers based on https://github.com/kexinhuang12345/CASTER

Note:
- covers only supervised training stage
- input dimensionality assumed to be correct and meaningful according to the assumptions of the algorithm
- TODO: inclusion of the unsupervised stage (would require a pipeline hook)
- TODO: inclusion of the BIOSNAP dataset from the paper
- TODO: data loading using the input processing from the paper
 
- [x] Unit tests provided for these changes
- [x] Documentation and docstrings added for these changes using the [sphinx style](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html)

## Changes 

* Implementation of the CASTER model with the custom loss function
* An example script for invoking the model (performance not meaningful because of the incomplete implementation of the model)
* A unit test method to check the output dimensionality

### CODE REVIEW

File: chemicalx/models/caster.py [hunk 1]
```diff
@@ -1,18 +1,176 @@
...
+    .. seealso:: This model was suggested in https://github.com/AstraZeneca/chemicalx/issues/17
 
-    .. [huang2020] Huang, K., *et al.* (2020). `CASTER: Predicting drug interactions
-       with chemical substructure representation <https://doi.org/10.1609/aaai.v34i01.5412>`_.
-       *AAAI 2020 - 34th AAAI Conference on Artificial Intelligence*, 702–709.
+    .. [huang2020] Huang, Kexin, *et al.* (2020). `CASTER: Predicting Drug Interactions with Chemical Substructure
+                    Representation <https://doi.org/10.1609/aaai.v34i01.5412>`_. *AAAI*, 34(1), 702-709.
     """
+
+    class CASTERSupervisedLoss(_Loss):
```
Comment (cthoyt) [inline chemicalx/models/caster.py:24]: please unnest

File: chemicalx/models/caster.py [hunk 2]
```diff
@@ -1,18 +1,176 @@
...
+                + self.lambda2 * torch.norm(dictionary_features_latent, p="fro") / batch_size
+            )
+
+            loss = loss_prediction + loss_reconstruction + loss_projection
+            return loss
+
+    def __init__(
+        self,
+        *,
+        drug_channels: int = 1722,
```
Comment (cthoyt) [inline chemicalx/models/caster.py:67]: this should be set based on dataset and not have a default

File: chemicalx/models/caster.py [hunk 3]
```diff
@@ -1,18 +1,176 @@
...
+
+        # decoder
+        self.decoder = torch.nn.Sequential(
+            torch.nn.Linear(encoder_output_channels, decoder_hidden_channels),
+            torch.nn.ReLU(True),
+            torch.nn.Linear(decoder_hidden_channels, drug_channels),
+        )
+
+        # predictor: eight layer NN
+        predictor_layers_dict = OrderedDict()
```
Comment (cthoyt) [inline chemicalx/models/caster.py:107]: this is very hard to think about. can you just use a list?

File: chemicalx/models/caster.py [hunk 4]
```diff
@@ -1,18 +1,176 @@
...
+        DTD_inv = torch.inverse(DTD + self.lambda3 * (torch.eye(self.drug_channels)))
+        DTD_inv_DT = torch.matmul(DTD_inv, dictionary_features_latent)
+        r = drug_pair_features_latent[:, None, :].matmul(DTD_inv_DT.transpose(2, 1)).squeeze(1)
+        return r
+
+    def forward(self, drug_pair_features: torch.FloatTensor) -> Tuple[torch.FloatTensor]:
+        """
+        Run a forward pass of the CASTER model
+
+        :param drug_pair_features (torch.FloatTensor): functional representation of each drug pair (see unpack method)
```
Comment (cthoyt) [inline chemicalx/models/caster.py:152]: better to make a named tuple class to document this kind of thing

**Comment by cthoyt** [comment]
@andriy-nikolov note I did some reorganization of the code in this PR. Make sure you pull before you begin to work on it again

**Review by benedekrozemberczki — APPROVED** [review]

**Comment by cthoyt** [comment]
I have several concerns with this PR, would have been nice to do a review first. Most importantly: why does it change the standard interface of the forward() function? I don't see where any of the other things it returns are used

**Comment by benedekrozemberczki** [comment]
The paper provides two types of training techniques - supervised and unsupervised. In the unsupervised setting you could use any type of drug pair dataset. This forward pass allows for both setups.
